File: php/api/objects/tracks.php
<?php
/*
 * Copyright (c) 2019 by Yann39.
 *
 * This file is part of Chachatte Team application.
 *
 * Chachatte Team is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Chachatte Team is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Chachatte Team. If not, see <http://www.gnu.org/licenses/>.
 */

class Track {

    // database connection and table name
    private $conn;
    private $table_name = "tracks";

    // object properties
    public $id;
    public $name;
    public $distance;
    public $lap_record;
    public $website;
    public $latitude;
    public $longitude;

    // constructor with $db as database connection
    public function __construct($db){
        $this->conn = $db;
    }

    // get all tracks
    function read() {

        // query to get all records
        $query = "SELECT n.id, n.name, n.distance, n.lap_record, n.website, n.latitude, n.longitude FROM " . $this->table_name . " n ORDER BY n.name";

        // prepare query statement
        $stmt = $this->conn->prepare($query);

        // execute query
        $stmt->execute();

        return $stmt;
    }

    // get a track given its id
    function readOne() {

        // query to get record corresponding to specified id
        $query = "SELECT n.id, n.name, n.distance, n.lap_record, n.website, n.latitude, n.longitude FROM " . $this->table_name . " n WHERE n.id = ? LIMIT 0,1";

        // prepare query statement
        $stmt = $this->conn->prepare($query);

        // bind id
        $stmt->bindParam(1, $this->id);

        // execute query
        $stmt->execute();

        // get retrieved row
        $row = $stmt->fetch(PDO::FETCH_ASSOC);

        // set values to object properties
        $this->name = $row['name'];
        $this->distance = $row['distance'];
        $this->lap_record = $row['lap_record'];
        $this->website = $row['website'];
        $this->latitude = $row['latitude'];
        $this->longitude = $row['longitude'];
    }

    // create track
    function create() {

        // query to insert record
        $query = "INSERT INTO " . $this->table_name . " SET name = :name, distance = :distance, lap_record = :lap_record, website = :website, latitude = :latitude, longitude = :longitude";

        // prepare query
        $stmt = $this->conn->prepare($query);

        // sanitize
        $this->name=htmlspecialchars(strip_tags($this->name));
        $this->distance=htmlspecialchars(strip_tags($this->distance));
        $this->lap_record=htmlspecialchars(strip_tags($this->lap_record));
        $this->website=htmlspecialchars(strip_tags($this->website));
        $this->latitude=htmlspecialchars(strip_tags($this->latitude));
        $this->longitude=htmlspecialchars(strip_tags($this->longitude));

        // bind values
        $stmt->bindParam(":name", $this->name);
        $stmt->bindParam(":distance", $this->distance);
        $stmt->bindParam(":lap_record", $this->lap_record);
        $stmt->bindParam(":website", $this->website);
        $stmt->bindParam(":latitude", $this->latitude);
        $stmt->bindParam(":longitude", $this->longitude);

        // execute query
        if ($stmt->execute()) {
            return $this->conn->lastInsertId();
        }

        return -1;
    }

    // update track
    function update() {

        // query to update record
        $query = "UPDATE " . $this->table_name . " SET name = :name, distance = :distance, lap_record = :lap_record, website = :website, latitude = :latitude, longitude = :longitude WHERE id = :id";

        // prepare query statement
        $stmt = $this->conn->prepare($query);

        // sanitize
        $this->id=htmlspecialchars(strip_tags($this->id));
        $this->name=htmlspecialchars(strip_tags($this->name));
        $this->distance=htmlspecialchars(strip_tags($this->distance));
        $this->lap_record=htmlspecialchars(strip_tags($this->lap_record));
        $this->website=htmlspecialchars(strip_tags($this->website));
        $this->latitude=htmlspecialchars(strip_tags($this->latitude));
        $this->longitude=htmlspecialchars(strip_tags($this->longitude));

        // bind new values
        $stmt->bindParam(':id', $this->id);
        $stmt->bindParam(":name", $this->name);
        $stmt->bindParam(":distance", $this->distance);
        $stmt->bindParam(":lap_record", $this->lap_record);
        $stmt->bindParam(":website", $this->website);
        $stmt->bindParam(":latitude", $this->latitude);
        $stmt->bindParam(":longitude", $this->longitude);

        // execute the query
        if ($stmt->execute()) {
            return true;
        }

        return false;
    }

    // delete track
    function delete() {

        // query to delete record
        $query = "DELETE FROM " . $this->table_name . " WHERE id = ?";

        // prepare query
        $stmt = $this->conn->prepare($query);

        // sanitize
        $this->id = htmlspecialchars(strip_tags($this->id));

        // bind id of record to delete
        $stmt->bindParam(1, $this->id);

        // execute query
        if ($stmt->execute()) {
            return true;
        }

        return false;
    }

    // search track
    function search($keywords){

        // query to search across all records
        $query = "SELECT n.name, n.distance, n.lap_record, n.website, n.latitude, n.longitude FROM " . $this->table_name . " n WHERE n.name LIKE ? ORDER BY n.name DESC";

        // prepare query statement
        $stmt = $this->conn->prepare($query);

        // sanitize
        $keywords = htmlspecialchars(strip_tags($keywords));
        $keywords = "%{$keywords}%";

        // bind values
        $stmt->bindParam(1, $keywords);

        // execute query
        $stmt->execute();

        return $stmt;
    }
}

File: lib/ui/tracks/tracks.dart
/*
 * Copyright (c) 2019 by Yann39.
 *
 * This file is part of Chachatte Team application.
 *
 * Chachatte Team is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Chachatte Team is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Chachatte Team. If not, see <http://www.gnu.org/licenses/>.
 */

import 'package:chachatte_team/models/track.dart';
import 'package:chachatte_team/providers/event_provider.dart';
import 'package:chachatte_team/providers/track_provider.dart';
import 'package:chachatte_team/ui/main/main_action_menu.dart';
import 'package:chachatte_team/ui/main/main_drawer.dart';
import 'package:chachatte_team/utils/custom_decorations.dart';
import 'package:chachatte_team/utils/date_utils.dart';
import 'package:chachatte_team/utils/strings.dart';
import 'package:chachatte_team/utils/track_utils.dart';
import 'package:chachatte_team/widgets/loading_content.dart';
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:provider/provider.dart';

/// We need stateful widget to keep search field value
class Tracks extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    return _TracksState();
  }
}

class _TracksState extends State<Tracks> {
  /// Build the search field
  TextField buildSearchField(TrackProvider _trackProvider) {
    return TextField(
      decoration: InputDecoration(
        filled: true,
        fillColor: Colors.blue[100],
        prefixIcon: Icon(Icons.search),
        hintText: AppString.tracksSearchHint,
      ),
      maxLines: 1,
      onChanged: (String text) {
        _trackProvider.searchTracks(text);
      },
    );
  }

  /// Method that launches the Track detail screen and awaits the result from Navigator.pop
  void _navigateToTrackDetailScreen(BuildContext context, Track track) async {
    /*Provider.of<EventProvider>(context, listen: false)
        .fetchTrackEvents(track.id);*/

    // Navigator.push returns a Future that will complete after we call Navigator.pop on the target screen
    final _result =
        await Navigator.pushNamed(context, '/trackDetail', arguments: track);

    // after the target screen returns a result, hide any previous snack bars and show the result
    if (_result != null) {
      ScaffoldMessenger.of(context)
        ..removeCurrentSnackBar()
        ..showSnackBar(SnackBar(content: Text("$_result")));
    }
  }

  Widget build(BuildContext context) {
    final _trackProvider = Provider.of<TrackProvider>(context, listen: true);

    return Scaffold(
      appBar: AppBar(
        title: Text(AppString.tabTracks),
        actions: <Widget>[MainActionMenu()],
      ),
      drawer: MainDrawer(),
      body: Column(
        children: <Widget>[
          buildSearchField(_trackProvider),
          Expanded(
            child: Container(
              decoration: CustomDecorations.mainContent,
              child: LoadingContent(
                emptyText: AppString.tracksNotFound,
                loadingStatus: _trackProvider.loadingStatus,
                child: GridView.builder(
                  padding: EdgeInsets.all(4.0),
                  gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                      crossAxisCount: MediaQuery.of(context).orientation ==
                              Orientation.portrait
                          ? 2
                          : 3,
                      crossAxisSpacing: 4,
                      mainAxisSpacing: 4,
                      childAspectRatio: 1),
                  itemCount: _trackProvider.tracks.length,
                  itemBuilder: (BuildContext context, int index) => InkWell(
                    onTap: () => _navigateToTrackDetailScreen(
                        context, _trackProvider.tracks[index]),
                    child: Card(
                      shape: RoundedRectangleBorder(
                          borderRadius: BorderRadius.circular(6.0)),
                      clipBehavior: Clip.antiAlias,
                      child: Container(
                        padding: EdgeInsets.all(8.0),
                        decoration: CustomDecorations.cardFull,
                        child: Column(
                          crossAxisAlignment: CrossAxisAlignment.center,
                          mainAxisAlignment: MainAxisAlignment.start,
                          children: <Widget>[
                            Padding(
                              padding: const EdgeInsets.only(bottom: 5.0),
                              child: Text(_trackProvider.tracks[index].name,
                                  style: TextStyle(color: Colors.white),
                                  textScaleFactor: 1.1),
                            ),
                            Divider(
                              height: 1.0,
                              color: Colors.white,
                            ),
                            Container(
                              height: 90.0,
                              padding: EdgeInsets.all(0),
                              child: TrackUtils.getTrackIcon(
                                  _trackProvider.tracks[index].name),
                            ),
                            Column(
                              crossAxisAlignment: CrossAxisAlignment.start,
                              mainAxisAlignment: MainAxisAlignment.start,
                              children: <Widget>[
                                Row(
                                  children: <Widget>[
                                    Icon(Icons.straighten,
                                        size: 13, color: Colors.white),
                                    SizedBox(width: 6.0),
                                    Text(
                                        "${AppString.length} : ${(_trackProvider.tracks[index].distance / 1000).toStringAsFixed(2)} km",
                                        style: TextStyle(color: Colors.white),
                                        textScaleFactor: 0.9),
                                  ],
                                ),
                                Row(
                                  children: <Widget>[
                                    Icon(Icons.timer,
                                        size: 13, color: Colors.white),
                                    SizedBox(width: 6.0),
                                    Text(
                                        "${AppString.record} : ${AppDateUtils.toLapTimeString(_trackProvider.tracks[index].lapRecord)}",
                                        style: TextStyle(color: Colors.white),
                                        textScaleFactor: 0.9),
                                  ],
                                ),
                              ],
                            ),
                          ],
                        ),
                      ),
                    ),
                  ),
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}
